### PR TITLE
Guided Tours: Find & show tours based on state selectors

### PR DESCRIPTION
**Pull request by mcsf**
Previously, GuidedTours was able to trigger only a single tour -- `main`. This was based solely on the presence of `tour=main` in the query arguments. This change introduces a selectors-based approach to triggering tours that is based on the current context and tour-specific requirements for that context. 

Duplicate of #6335, but based on `master` this time.
Product of exploration in #6146. See that PR for context.

### Before merging
- [x] Tests
- [x] For consistency, settle on name: `tour` or `tourName` (punted)
- [x] Assimilate `relevantFeatures` into config, or leave that to next PR? (next PR)
- [x] Don't actually show dummy tours in production
- [x] Misc. cleanup, squash
- [x] Consider only saving _certain_ `ROUTE_SET` actions in state, for performance (punted, later when we need it)

### Testing

- get into the guidedTours abtest***, create a new user account, and make sure after signup you are redirected to a `?tour=main` URL
- make sure the tour works fine
- make sure you can visit other parts of Calypso without throwing an error (especially try clicking on themes)
- make sure you can still summon the tour by appending the `?tour=main` bit

*** `localStorage.setItem( 'ABTests', '{"guidedTours_20160603":"guided"}' );`

Test live: https://calypso.live/?branch=add/guided-tours-log-selectors-rebased

### CODE REVIEW

**Comment by mcsf** [comment]
One thing I noticed worries me.

The following breaks the tests, **as it should**:

```diff
diff --git a/client/state/ui/guided-tours/selectors.js b/client/state/ui/guided-tours/selectors.js
index 74af0d5..e7e3d1b 100644
--- a/client/state/ui/guided-tours/selectors.js
+++ b/client/state/ui/guided-tours/selectors.js
@@ -92,7 +92,7 @@ export const findEligibleTour = createSelector(
 			return context( state );
 		} );
 	},
-	[ getActionLog, getToursHistory ]
+	getActionLog
 );
 
 /**
```

because the `findEligibleTour` selector gets called multiple times in that test suite and needs to know that it depends on `getToursHistory` to work properly.

**However**, the following **does not** break the tests, when I believe it should be functionally equivalent to the previous diff:

```diff
diff --git a/client/state/ui/guided-tours/selectors.js b/client/state/ui/guided-tours/selectors.js
index 74af0d5..2c4059f 100644
--- a/client/state/ui/guided-tours/selectors.js
+++ b/client/state/ui/guided-tours/selectors.js
@@ -92,7 +92,7 @@ export const findEligibleTour = createSelector(
 			return context( state );
 		} );
 	},
-	[ getActionLog, getToursHistory ]
+	[ getActionLog ]
 );
 
 /**
```

@aduth, since we've been discussing selectors and have just merged #6391, I'm cc'ing you for a sanity check. It's getting late here.

**Comment by aduth** [comment]
Quite possibly because memoization cache is not reset between tests. [See posts tests as an example](https://github.com/Automattic/wp-calypso/blob/f248c9aecc968dc01321d6a10230cda1cf22470b/client/state/posts/test/selectors.js#L29-L34).

**Comment by mcsf** [comment]
Yes, but I expect memoization to be working. However, I'm surprised that `getActionLog( state )` is treated differently from `[ getActionLog( state ) ]`.

**Comment by aduth** [comment]
Hmm, equally confusing, I find all tests passing when applying your first diff, but resetting selectors between tests.

```diff
diff --git a/client/state/ui/guided-tours/selectors.js b/client/state/ui/guided-tours/selectors.js
index 74af0d5..e279a62 100644
--- a/client/state/ui/guided-tours/selectors.js
+++ b/client/state/ui/guided-tours/selectors.js
@@ -45,7 +45,7 @@ const relevantFeatures = [
        },
 ];
 
-const getToursFromFeaturesReached = createSelector(
+export const getToursFromFeaturesReached = createSelector(
        state => reverse( uniq(
                getActionLog( state )
                        .filter( ( { type } ) => type === ROUTE_SET )
@@ -63,7 +63,7 @@ const getToursFromFeaturesReached = createSelector(
        getActionLog
 );
 
-const getToursSeen = createSelector(
+export const getToursSeen = createSelector(
        state => uniq(
                getToursHistory( state )
                        .map( ( { tourName } ) => tourName )
@@ -92,7 +92,7 @@ export const findEligibleTour = createSelector(
                        return context( state );
                } );
        },
-       [ getActionLog, getToursHistory ]
+       getActionLog
 );
 
 /**
diff --git a/client/state/ui/guided-tours/test/selectors.js b/client/state/ui/guided-tours/test/selectors.js
index 94b4557..bb1b0c4 100644
--- a/client/state/ui/guided-tours/test/selectors.js
+++ b/client/state/ui/guided-tours/test/selectors.js
@@ -7,12 +7,21 @@ import { expect } from 'chai';
  * Internal dependencies
  */
 import {
+       getToursFromFeaturesReached,
+       getToursSeen,
        getGuidedTourState,
        findEligibleTour,
 } from '../selectors';
 import guidedToursConfig from 'layout/guided-tours/config';
 
 describe( 'selectors', () => {
+       beforeEach( () => {
+               getToursFromFeaturesReached.memoizedSelector.cache.clear();
+               getToursSeen.memoizedSelector.cache.clear();
+               getGuidedTourState.memoizedSelector.cache.clear();
+               findEligibleTour.memoizedSelector.cache.clear();
+       } );
+
        describe( '#getGuidedTourState()', () => {
                it( 'should return an empty object if no state is present', () => {
                        const tourState = getGuidedTourState( {
```

**Comment by aduth** [comment]
Aside: I contemplated the idea that we could check in `lib/create-selector` whether the current environment is test and disabling cache, if it becomes a pain to deal with memoization reset in tests (cc @blowery ).

**Comment by mcsf** [comment]
> Hmm, equally confusing, I find all tests passing when applying your first diff, but resetting selectors cache between tests.

That one actually sounds correct, because — even though we're setting the wrong dependencies by only mentioning `getActionLog` — by busting the cache we make sure that the selectors recompute everything. Since the state and the implementations are good enough, the tests then pass.

**Comment by mcsf** [comment]
> Aside: I contemplated the idea that we could check in lib/create-selector whether the current environment is test and disabling cache, if it becomes a pain to deal with memoization reset in tests (cc @blowery ).

That's an interesting idea. I bet there's a higher probability of a bug not being spotted due to a selector's cache not being cleared than the opposite — even though, in this particular instance, I spotted the need to add the `getToursHistory` dependency thanks to the cached result not making sense.

**Comment by mcsf** [comment]
Rebased against `master`. An external fix, #6435, improves performance when browsing `/themes`.

File: client/state/notices/actions.js [hunk 1]
```diff
@@ -36,10 +36,11 @@ export function createNotice( status, text, options = {} ) {
 	};
 }
 
-export function setRoute( path ) {
+export function setRoute( path, query = {} ) {
```
Comment (aduth) [inline client/state/notices/actions.js]: In #6306 (to be merged shortly), this action creator is moved to `state/ui/actions.js`

File: client/state/ui/action-log/reducer.js [hunk 2]
```diff
@@ -12,22 +11,33 @@ import {
 	GUIDED_TOUR_UPDATE,
 	THEMES_RECEIVE,
 	PREVIEW_IS_SHOWING,
-	ROUTE_SET,
 } from 'state/action-types';
 
-const isRelevantActionType = includes.bind( null, [
+const isDesignRoute = ( { path } ) => 0 === path.indexOf( '/design' );
```
Comment (ehg) [inline client/state/ui/action-log/reducer.js]: This seems a bit too specific to have in here `actionLog` selectors.

File: client/state/ui/guided-tours/actions.js [hunk 3]
```diff
@@ -21,14 +21,16 @@ import { getPreference } from 'state/preferences/selectors';
...
+export function showGuidedTour( { shouldShow, shouldDelay = false, tour } ) {
 	const showAction = {
 		type: GUIDED_TOUR_SHOW,
 		shouldShow,
 		shouldDelay,
 		tour,
 	};
 
+	// TODO(mcsf): track this *somewhere*, now that we'll stop relying on this
+	// action for launching tours
```
Comment (ehg) [inline client/state/ui/guided-tours/actions.js]: Any idea how we'd do this? Maybe a `isFirstShow` in the GT selector, then `<GuidedTours>` can fire off the action.

File: client/state/ui/guided-tours/selectors.js [hunk 4]
```diff
@@ -2,16 +2,165 @@
  * External dependencies
  */
 import get from 'lodash/get';
```
Comment (ehg) [inline client/state/ui/guided-tours/selectors.js]: We can do this now :D https://github.com/Automattic/wp-calypso/pull/6539/files#diff-4aaad6137f1bc91a0f39e5b5e3ddb478R6

File: client/state/ui/action-log/reducer.js (hunk 2 above)
Comment (mcsf) [inline client/state/ui/action-log/reducer.js]: That's true, but I'm not sure how to solve it _right now_. I'm leaning towards leaving it here for now and solving it in another PR.

Comment (ehg) [inline client/state/ui/action-log/reducer.js]: OK. Will it affect `FirstView` stuff?

File: client/state/ui/guided-tours/actions.js (hunk 3 above)
Comment (mcsf) [inline client/state/ui/guided-tours/actions.js]: Discussing this with @lsinger — there would be more than one way. The bare minimum would be to _cheat_ and still manually dispatch something in `boot` for the purpose of analytics, which something more sophisticated is conceived.

For instance, if we were to follow the route of implementing some assistant GT middleware, that could be a piece responsible for analytics. A less sophisticated alternative would be to let the view layer (perhaps in `componentDidUpdate`) look for cases where `stepName === 'init'` and decide to dispatch a `calypso_guided_tours_show` event — caveat would be making sure we don't dispatch duplicate events if something unexpected happens with the props.

Comment (ehg) [inline client/state/ui/guided-tours/actions.js]: > The bare minimum would be to cheat and still manually dispatch something in boot for the purpose of analytics, which something more sophisticated is conceived.

As long as it's equivalent to what we have now, I don't mind that :) I guess we can sort it out properly in another PR, before the themes tour goes live.

File: client/state/ui/action-log/reducer.js (hunk 2 above)
Comment (mcsf) [inline client/state/ui/action-log/reducer.js]: Good question. So we either disable this optimization, or if we really care we extend it to also log Stats routes.

File: client/state/ui/guided-tours/selectors.js [hunk 5]
```diff
@@ -2,16 +2,165 @@
...
 import createSelector from 'lib/create-selector';
 import guidedToursConfig from 'layout/guided-tours/config';
 
 const getToursConfig = memoize( ( tour ) => guidedToursConfig.get( tour ) );
+const getToursHistory = state => getPreference( state, 'guided-tours-history' );
+
+/*
+ * This would/will be part of the existing config for tours.
+ */
+const relevantFeatures = [
```
Comment (mcsf) [inline client/state/ui/guided-tours/selectors.js]: If that's OK with you, I'd leave that to another PR to get this one moving.

File: client/state/ui/action-log/reducer.js (hunk 2 above)
Comment (ehg) [inline client/state/ui/action-log/reducer.js]: I would disable it. If we have perf problems, we can revisit.

**Comment by mcsf** [comment]
Pushed a large rebase, with a lot of the feedback addressed right into pre-existing (repackaged) commits. I'll do some more squashing soon in a way that keeps the history sound.

@ehg: maybe wait until the second rebase before updating #6472 :)

**Comment by mcsf** [comment]
Thanks for tackling those outstanding items, @lsinger! The logic after the removal of the dummy tours seems good.

**Comment by ehg** [comment]
We should test whether the `shouldDelay` stuff still works with checkout before we consider merging :)

**Comment by lsinger** [comment]
> We should test whether the shouldDelay stuff still works with checkout before we consider merging :)

Hgrmpf.

screen shot 2016-07-15 at 10 30 59 am

**Comment by ehg** [comment]
Rebased.

**Comment by ehg** [comment]
Let's merge this.

**Comment by ockham** [comment]
Stumbled across the `initial` and `current` qargs reducers/selectors while looking at #19036. This adds considerable complexity and confusion for me TBH -- I don't know what initial vs current means for qargs without having any context, and when working with them elsewhere, I don't know how to deal with that distinction.

Do we really need both?

**Comment by mcsf** [comment]
tl;dr: `initial` means _these are the qargs with which Calypso booted_, `current` means _these are the actual latest qargs, as seen in `location`._

IIRC, the following would happen:

1. Follow a link to `//calypso/?tour=main`
2. a. Either during init, Calypso adjusts the route and in that process the qarg is lost,
2. b. or the tour starts, but as the user follows tour instructions and navigates, qarg is lost
3. Loss of qarg messes up decision process in state selectors to keep on with the tour

This led to the need for `initial`. I can't remember if `current` was ever needed _for us_, or if it was introduced in opposition to `initial`. The rationale would be that, if there were only a single reducer `qarg`, a dev would expect it to reflect the latest browser state, not the initial one.

**Comment by lsinger** [comment]
I noticed today that triggering tours using `?tour=tourName` doesn't work anymore — well, it does work, but only once until you clear your redux state. 

Something @markryall found:

> the first value of that query string is stored in the redux state and subsequent values are ignored

Does anyone have any clue how this might have come to be?

**Comment by ockham** [comment]
@lsinger Could this be related? #20255

**Comment by lsinger** [comment]
Yes, thanks @ockham. Document here: https://github.com/Automattic/wp-calypso/issues/20553#issuecomment-351370142
